pkg/providers/instance/vmextension.go

```diff
@@ -0,0 +1,115 @@
+package instance
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute"
+	"github.com/samber/lo"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	vmExtensionType = "Microsoft.Compute/virtualMachines/extensions"
+)
+
+const (
+	cseNameWindows      = "windows-cse-agent-karpenter"
+	cseTypeWindows      = "CustomScriptExtension"
+	csePublisherWindows = "Microsoft.Compute"
+	cseVersionWindows   = "1.10"
+	cseNameLinux        = "cse-agent-karpenter"
+	cseTypeLinux        = "CustomScript"
+	csePublisherLinux   = "Microsoft.Azure.Extensions"
+	cseVersionLinux     = "2.0"
+)
+
+const (
+	aksIdentifyingExtensionName      = "computeAksLinuxBilling"
+	aksIdentifyingExtensionPublisher = "Microsoft.AKS"
+	aksIdentifyingExtensionTypeLinux = "Compute.AKS.Linux.Billing"
+)
+
+// createAKSIdentifyingExtension attaches a VM extension to identify that this VM participates in an AKS cluster
+func (p *DefaultProvider) createAKSIdentifyingExtension(ctx context.Context, vmName string) (err error) {
```

Comment:
style opinion: I'm _really_ not a fan of this "split a struct across N files" pattern.

The code here is too coupled to code where the DefaultProvider is defined to really be standalone, and IMO all this serves is to hide when a struct is too big - i.e. you look at a module and see 7 nicely sized files and then realize "wait these are all just methods on a single uber-struct that does everything".

IMO file complexity/structure should match (as closely as possible) code complexity/structure. If you think this stuff deserves its own file (I don't disagree) then it also probably deserves its own structure. OTOH, if we don't want to do that, then IMO just put it in the instance.go file so we're honest about how much there is there.

Reply:
For me the biggest benefit is when we have two people working on a change in the instance.go we don't have conflicts. Also you immediately know where to go to find the drift code in cloudprovider for example. 

Definitions are in vmextensions whereas invocations are in instance.go(AKA MEGA STRUCT). I was tempted to do what you are mentioning, but then I feel i would ahve to for both nic and vm.